### ISSUE TITLE
Link backend resources with frontend

### ISSUE THREAD

**Issue by DamonGreenhalgh** [uoa-compsci399-s2-2022/spatial-skills-web-app#16]
## Feature Description
Link the back-end systems like question data to the front-end to be displayed within the test page.

## User Story
"As a developer I would like to use the MVC pattern to deliver this product"

## MVP
Be able to fetch question data from the front-end, this includes the image, text and potential answer images (where images are just the links to the CDN).

**Comment by wukelang** [comment]
Solved the CORS problem yesterday by using [https://www.npmjs.com/package/cors](url). I just didn't import it correctly since I didn't realise `app.js` was on ES6.

I have updated the `app.js` file and am now able to retrieve the json from our localhost endpoints
![image](https://user-images.githubusercontent.com/29995754/189019149-4f47271e-5ddb-4381-9edc-a91096a9a9a8.png)

**Comment by wukelang** [comment]
### Progress
Changed the testing page so that it works with JSON instead of the local files.
![image](https://user-images.githubusercontent.com/29995754/189114846-e9c924ca-5367-4660-a376-b487755f5b71.png) 
It currently uses raw JSON copy pasted from an the api/question/all endpoint. 

Format of an answer is now `[question id, answer id]` to correspond with the studentanswer schema in the DB. 

### Difficulties

- Retrieving JSON data from the endpoints and storing it in a variable
  - I could only 'store' the JSON data by updating a state variable in the test component with the retrieved JSON.
  - However, this caused had some async problems as the state only changed after the component fully renders, so other variables dependent on the JSON would cause the app to crash.
  - Storing the variable outside the component was difficult as both Fetch API and Axios return `Promise` objects when accessing the endpoint, instead of an array which we can actually use.

- Accessing Google Drive assets too many times (~15 times) makes them block your IP temporarily (403 Forbidden)
  - Actually accessing the files seem to be okay, although it takes some time to load (~2 seconds)
  - ![image](https://user-images.githubusercontent.com/29995754/189119069-553d3e85-2618-440e-a383-004cc739db31.png)

- DB stuff
  - I'm thinking about how we should get the questions from the test object. Seems like we would need to call the question/[id] endpoint for every question to get its data. Is there a better way to do this?
  - Question types; multichoice, entry, memory game question, select all (multiple correct answers). I think we can focus on multichoice and memory for now and slowly expand scope later


Other stuff to add: check if the user has selected an answer (answer is not null) by using a confirmation alert when they click the next button

**Comment by ryanla-bme** [comment]
## Response to DB and server-side issues

### Slow and blocked access to Google Drive Assets.

**Proposed solution:**

I will change the file storage to be in MongoDB natively, although this will come with storage space limitations (512 MB) and will require some data model changes. But this would work better if we plan to implement a question upload feature later as well. I can also explore other CDN options such as Amazon CloudFront or Google Cloud CDN, but these will most likely have costs associated with them.

### Getting questions from test object

**Proposed solution:**

I will implement another endpoint .../api/test/getQuestions/:tId where you will get an array of questions that are in the test specified by the test Id (tId).

**Comment by wukelang** [comment]
### Progress
Got the Test component to retrieve the api endpoint asynchronously using useEffect() (https://reactjs.org/docs/faq-ajax.html). 
I'm currently just using the `api/question/all` endpoint to get all questions, but I'll look into using a Test endpoint soon.
`  useEffect(() => {
    fetch('http://localhost:3001/api/question/all')
    .then(res => res.json())
    .then(
      (result) => {
        setQuestionBank(result);
        setUserAnswers(Array.apply(null, Array(result.length)));
        setIsLoaded(true);
      }`

### G Drive
> I will change the file storage to be in MongoDB natively, although this will come with storage space limitations (512 MB) and will require some data model changes. But this would work better if we plan to implement a question upload feature later as well. I can also explore other CDN options such as Amazon CloudFront or Google Cloud CDN, but these will most likely have costs associated with them.

I think when someone normally uses this app they won't trigger the 403 (unless they spam refresh on a test), but it is something to consider.

### Endpoints
Another endpoint I'd like if possible is a "get question time array" for a test. Currently I would need to use a loop to go through each question and extract their time value in the test object.

**Comment by ryanla-bme** [comment]
## Progress on DB and server-side

### Database

I have decided to use AWS S3 as our CDN as this is provided to us. Hopefully this will solve the problems when accessing images. However, currently we have not been given access to the AWS SDK and cannot implement file upload without this. I have asked for access and waiting for a response.

### Endpoints 

I have added an endpoint you can access from ~~GET~~ POST .../api/test/getquestions

It returns questions from the test and also takes in parameter whether to be shuffled. Included in the response object is an array with the times of the questions. These will correspond to the times specified in the tId for the questions in the order the questions where shuffled (if shuffled).

Example:

Request body:

```json
{
    "tId": "6319abdf2d143b5bfa3de54a",
    "shuffle": true
}
```

Response body: 

```json
{
    "questions": [
        {
            "id": "631701580df40a96f0a7f8b7",
            "title": "Perception 1",
            "description": "Which of the following cube can you make with these four pieces?",
            "image": "https://team-22-question-bucket.s3.ap-southeast-2.amazonaws.com/PERCEPTION/perception1/question.png",
            "answer": [
                {
                    "id": "631701580df40a96f0a7f8b8",
                    "image": "https://team-22-question-bucket.s3.ap-southeast-2.amazonaws.com/PERCEPTION/perception1/a.png"
                },
                {
                    "id": "631701580df40a96f0a7f8b9",
                    "image": "https://team-22-question-bucket.s3.ap-southeast-2.amazonaws.com/PERCEPTION/perception1/b.png"
                },
                {
                    "id": "631701580df40a96f0a7f8ba",
                    "image": "https://team-22-question-bucket.s3.ap-southeast-2.amazonaws.com/PERCEPTION/perception1/c.png"
                },
                {
                    "id": "631701580df40a96f0a7f8bb",
                    "image": "https://team-22-question-bucket.s3.ap-southeast-2.amazonaws.com/PERCEPTION/perception1/d.png"
                }
            ],
            "category": "Visual perception"
        },
        {
            "id": "6316ff1b0df40a96f0a7f8aa",
            "title": "Rotation 2",
            "description": "Which figure is identical to the first?",
            "image": "https://team-22-question-bucket.s3.ap-southeast-2.amazonaws.com/ROTATION/rotation2/question.png",
            "answer": [
                {
                    "id": "6316ff1b0df40a96f0a7f8ab",
                    "image": "https://team-22-question-bucket.s3.ap-southeast-2.amazonaws.com/ROTATION/rotation2/a.png"
                },
                {
                    "id": "6316ff1b0df40a96f0a7f8ac",
                    "image": "https://team-22-question-bucket.s3.ap-southeast-2.amazonaws.com/ROTATION/rotation2/b.png"
                },
                {
                    "id": "6316ff1b0df40a96f0a7f8ad",
                    "image": "https://team-22-question-bucket.s3.ap-southeast-2.amazonaws.com/ROTATION/rotation2/c.png"
                },
                {
                    "id": "6316ff1b0df40a96f0a7f8ae",
                    "image": "https://team-22-question-bucket.s3.ap-southeast-2.amazonaws.com/ROTATION/rotation2/d.png"
                }
            ],
            "category": "Mental rotation"
        },
        {
            "id": "6316fb5c0df40a96f0a7f898",
            "title": "Visualisation 1",
            "description": "Which one of the following will be the final pattern after these actions occur?",
            "image": "https://team-22-question-bucket.s3.ap-southeast-2.amazonaws.com/VISUALISATION/visualisation1/question.png",
            "answer": [
                {
                    "id": "6316fb5c0df40a96f0a7f899",
                    "image": "https://team-22-question-bucket.s3.ap-southeast-2.amazonaws.com/VISUALISATION/visualisation1/a.png"
                },
                {
                    "id": "6316fb5c0df40a96f0a7f89a",
                    "image": "https://team-22-question-bucket.s3.ap-southeast-2.amazonaws.com/VISUALISATION/visualisation1/b.png"
                },
                {
                    "id": "6316fb5c0df40a96f0a7f89b",
                    "image": "https://team-22-question-bucket.s3.ap-southeast-2.amazonaws.com/VISUALISATION/visualisation1/c.png"
                },
                {
                    "id": "6316fb5c0df40a96f0a7f89c",
                    "image": "https://team-22-question-bucket.s3.ap-southeast-2.amazonaws.com/VISUALISATION/visualisation1/d.png"
                }
            ],
            "category": "Spatial visualisation"
        },
        {
            "id": "6316fdf10df40a96f0a7f8a4",
            "title": "Rotation 1",
            "description": "Which figure is identical to the first?",
            "image": "https://team-22-question-bucket.s3.ap-southeast-2.amazonaws.com/ROTATION/rotation1/question.png",
            "answer": [
                {
                    "id": "6316fdf10df40a96f0a7f8a5",
                    "image": "https://team-22-question-bucket.s3.ap-southeast-2.amazonaws.com/ROTATION/rotation1/a.png"
                },
                {
                    "id": "6316fdf10df40a96f0a7f8a6",
                    "image": "https://team-22-question-bucket.s3.ap-southeast-2.amazonaws.com/ROTATION/rotation1/b.png"
                },
                {
                    "id": "6316fdf10df40a96f0a7f8a7",
                    "image": "https://team-22-question-bucket.s3.ap-southeast-2.amazonaws.com/ROTATION/rotation1/c.png"
                },
                {
                    "id": "6316fdf10df40a96f0a7f8a8",
                    "image": "https://team-22-question-bucket.s3.ap-southeast-2.amazonaws.com/ROTATION/rotation1/d.png"
                }
            ],
            "category": "Mental rotation"
        },
        {
            "id": "6316fce60df40a96f0a7f89e",
            "title": "Visualisation 2",
            "description": "Which one of the following will be the final pattern after these actions occur?",
            "image": "https://team-22-question-bucket.s3.ap-southeast-2.amazonaws.com/VISUALISATION/visualisation2/question.png",
            "answer": [
                {
                    "id": "6316fce60df40a96f0a7f89f",
                    "image": "https://team-22-question-bucket.s3.ap-southeast-2.amazonaws.com/VISUALISATION/visualisation2/a.png"
                },
                {
                    "id": "6316fce60df40a96f0a7f8a0",
                    "image": "https://team-22-question-bucket.s3.ap-southeast-2.amazonaws.com/VISUALISATION/visualisation2/b.png"
                },
                {
                    "id": "6316fce60df40a96f0a7f8a1",
                    "image": "https://team-22-question-bucket.s3.ap-southeast-2.amazonaws.com/VISUALISATION/visualisation2/c.png"
                },
                {
                    "id": "6316fce60df40a96f0a7f8a2",
                    "image": "https://team-22-question-bucket.s3.ap-southeast-2.amazonaws.com/VISUALISATION/visualisation2/d.png"
                }
            ],
            "category": "Spatial visualisation"
        }
    ],
    "times": [
        7,
        11,
        10,
        12,
        8
    ]
}
```

**Comment by wukelang** [comment]
MVP for this issue is likely finished. 
The test page is able to get its questions from a test object in the database.

### Stuff I want to work on before merging

- [x] Error checking in api endpoints (e.g server doesn't crash for invalid ID)
- [x] Next button checks if the user has selected an answer / wants to submit the test.
- [x] A simple 'results' screen that shows the program can mark the students answers after the test finishes
- [x] Implement sliding timer bar

### Things to consider
- Refreshing the test page causes the entire test to restart. To prevent this, we would need a server side counter for current question and time, although I think this is really difficult to do
- How are we going to implement the memory games into the test
- We should make the time for each question in the database longer (~2 minutes).
- ~~Maybe make a file that has functions to call specific endpoints with axios (instead of hard coding it everywhere)~~ This only works for await calls I think
